src/meta/src/hummock/compaction/picker/compaction_task_validator.rs

```diff
@@ -0,0 +1,217 @@
+            max_level_size = std::cmp::max(max_level_size, level_select_size);
+
+            total_file_size += level_select_size;
+            total_file_count += select_level.table_infos.len();
+        }
+
+        // This limitation would keep our write-amplification no more than
+        // ln(max_compaction_bytes/flush_level_bytes) /
+        // ln(self.config.level0_sub_level_compact_level_count/2) Here we only use half
+        // of level0_sub_level_compact_level_count just for convenient.
+        let is_write_amp_large =
+            max_level_size * self.config.level0_sub_level_compact_level_count as u64 / 2
+                >= total_file_size;
+
+        if is_write_amp_large
+            && total_file_count < self.config.level0_max_compact_file_number as usize
+        {
+            stats.skip_by_write_amp_limit += 1;
+            return false;
+        }
+
+        if input.input_levels.len() < intra_sub_level_compact_level_count
+            && total_file_count < self.config.level0_max_compact_file_number as usize
+        {
+            stats.skip_by_count_limit += 1;
+            return false;
+        }
+
+        true
+    }
+
+    fn valid_base_level_compaction(
+        &self,
+        input: &CompactionInput,
+        stats: &mut LocalPickerStatistic,
+    ) -> bool {
+        let target_level_size: u64 = input
+            .input_levels
+            .last()
+            .as_ref()
+            .unwrap()
+            .table_infos
+            .iter()
+            .map(|sst| sst.file_size)
+            .sum();
+
+        let select_file_size: u64 = input
+            .input_levels
+            .iter()
+            .map(|level| {
+                level
+                    .table_infos
+                    .iter()
+                    .map(|sst| sst.file_size)
+                    .sum::<u64>()
+            })
+            .sum::<u64>()
+            - target_level_size;
+
+        // The size of target level may be too large, we shall skip this compact task and wait
+        //  the data in base level compact to lower level.
+        if target_level_size > self.config.max_compaction_bytes {
+            stats.skip_by_count_limit += 1;
+            return false;
+        }
+
+        if select_file_size < target_level_size {
+            stats.skip_by_write_amp_limit += 1;
+            return false;
+        }
+
+        true
+    }
+
+    pub fn valid_compact_task(
+        &self,
+        input: &CompactionInput,
+        picker_type: CompactionTaskOptimizeRule,
+        stats: &mut LocalPickerStatistic,
+    ) -> bool {
+        if !self.optimize_rules.contains(&picker_type) {
+            return true;
+        }
+
+        match picker_type {
+            CompactionTaskOptimizeRule::Tier => self.valid_tier_compaction(input, stats),
+            CompactionTaskOptimizeRule::Intra => self.valid_intra_compaction(input, stats),
+            CompactionTaskOptimizeRule::ToBase => self.valid_base_level_compaction(input, stats),
```

Comment:
It seems that different types of the rules share the same input parameters for validation. In such case, is it better structure the code by having a trait and letting each rule implement the trait on its own?
```
trait CompactionTaskValidationRule {
 fn validate(&self,  input: &CompactionInput, stats: &mut LocalPickerStatistic) -> bool
}

struct TierCompactionTaskValidationRule {
...
}

impl CompactionTaskValidationRule  for TierCompactionTaskValidationRule  {
 fn validate(&self,  input: &CompactionInput, stats: &mut LocalPickerStatistic) -> bool {
    ...
 }
}

...
```

I don't have a strong opinion on this though.